### DISCUSSION TITLE
[ELOQUENT] whereHas inside where

### DISCUSSION THREAD

**Discussion by fzn0x** [laravel/framework#38333] [Ideas] [answered]
Hello, maybe the title will make you confused because usually `where` is used inside `whereHas`, but there are some usecases where we have to make sure that the relation is available on certain data conditions.

e.g :
we have relation from `lesson` table to `classrooms` table, **but** only few `classrooms` was available for that `lesson` and the conditions used are in the `is_private` column.

```php
// filter something for private lesson only.
// $query is from `lesson` table.
$query->where('is_private', true, function ($subQuery) {
    // make sure only pick data that has classrooms for private lesson only.
    $subQuery->has('classrooms');

   // or using whereHas example
   $subQuery->whereHas('classrooms', function ($deptQuery) {
      $deptQuery->where('course_id', $course->id);
   });
});
```

i don't know how to achieve this more efficient with eloquent or raw query, but if you want to provide the example with the existing laravel features, I am very grateful you had help.

to make this easier to understand : **where if** but also check for relations availability.

```php
ModelName::whereRaw('IF (`is_private` = true, CHECK_RELATIONSHIP)->get();
```

**Comment by hotmeteor** [comment, marked as answer]
I think you just want to chain the two wheres, since there's no `or` involved.
```php
// using has
$query->where('is_private', true)->has('classrooms');
```
```php
// using whereHas
$query->where('is_private', true)->whereHas('classrooms', function ($deptQuery) {
      $deptQuery->where('course_id', $course->id);
});
```
